[REVIEW] Fix Padding Related Bugs:  `Crossfit`

VibhuJawa opened the pull request:
This PR plans to fix: 

- https://github.com/rapidsai/crossfit/issues/59
- https://github.com/rapidsai/crossfit/issues/65

TODO:

- [x] Add tests
- [x]  Verify fix for (https://github.com/rapidsai/crossfit/issues/65), @ryantwolf , Can probably help

sarahyurick reviewed (changes requested):
Could you please add an example test for `padding_side="right"` vs `padding_side="left"`?

VibhuJawa commented:
> Could you please add an example test for `padding_side="right"` vs `padding_side="left"`?

Added

ryantwolf commented on crossfit/backend/torch/hf/model.py:
```diff
@@ -81,11 +96,13 @@ def fit_memory_estimate_curve(self, model=None):
         max_seq = self.max_seq_length()
-        for batch_size in tqdm(range(2048, 0, -256)):
+        for batch_size in tqdm(
+            range(self.end_batch_size, self.start_batch_size, -self.batch_size_increment)
```
Might be best to do
```suggestion
            range(self.end_batch_size, self.start_batch_size - 1, -self.batch_size_increment)
```

Otherwise with `self.start_batch_size = 1` you won't actually hit the batch size of 1

VibhuJawa replied:
Resolved by: https://github.com/rapidsai/crossfit/pull/66/commits/bbdc4f2dec05cd31168bb8e0b2e9ad62ec586e54

ryantwolf commented on crossfit/backend/torch/hf/model.py:
```diff
@@ -81,11 +96,13 @@ def fit_memory_estimate_curve(self, model=None):
+            range(self.end_batch_size, self.start_batch_size, -self.batch_size_increment)
+        ):
             if batch_size <= 0:
                 continue
```
Is there a reason you start at the largest batch size first and then go down? In my experience with llama 7B, a lot of time was wasted on the larger batch sizes and larger sequence lengths. It would safe a bit of time if you could stop early. In the inner loop, you could stop when the sequence length is too long given the current batch size and it triggers an OOM. For the outer loop, you could stop when all sequence lengths at the current batch size triggered an OOM.

It's a minor thing compared to the time it takes to run the models on large datasets obviously, but it was annoying when repeatedly debugging with a small dataset.

VibhuJawa replied:
I agree, it makes sense to go from small to large as that allows early stopping.

VibhuJawa replied:
Resolved by: https://github.com/rapidsai/crossfit/pull/66/commits/bbdc4f2dec05cd31168bb8e0b2e9ad62ec586e54

ryantwolf commented on crossfit/backend/torch/loader.py:
```diff
@@ -66,7 +75,11 @@ def __next__(self):
         batch = {key: val[self.current_idx : end] for key, val in self.tensor_dict.items()}
         if self.max_seq_len is not None:
-            batch = {key: val[:, : self.max_seq_len] for key, val in batch.items()}
+            # TODO: Check this
```
Not sure if the right padding has been checked, but the left padding seems to work fine.

VibhuJawa replied:
Forgot to remove that TODO:

ryantwolf replied:
It's still there if it matters to you.

ryantwolf commented on crossfit/backend/torch/loader.py:
```diff
@@ -138,8 +159,17 @@ def __next__(self):
-                clip_len = min(max(_tokens[start], _tokens[end - 1]), self.model.max_seq_length())
-                batch = {key: val[:, :clip_len] for key, val in batch.items()}
+                # TODO: Fix max_length
+                if self.max_seq_len is None:
```
What is left to do here? Also, shouldn't `self.max_seq_len` always be set here since you set it in the constructor, or am I missing an edgecase?

VibhuJawa replied:
It should be but was not being set and i could not figure out why .

ryantwolf replied:
I figured it out. You are calling `super().__init__(frame, initial_batch_size, progress_bar=progress_bar)`on line 130 which resets it to None after you set it on line 122. I'd just remove the assignment on line 122 and pass `self.model.max_seq_length()` into the super constructor

VibhuJawa replied:
That was it, fixed it.

ryantwolf reviewed (commented):
Just some minor comments, thanks for fixing this!

VibhuJawa commented:
Resolved comments @ryantwolf

VibhuJawa commented:
Verified same results before and after memory changes

```python3
import os
import numpy as np
import joblib
from sklearn.linear_model import LinearRegression

new_model_dir = "/home/nfs/vjawa/.cf/memory" 
old_model_dir = "/home/nfs/vjawa/bkp_memory_curves"

model_name = "microsoft/deberta-v3-base"
#model_name = "sentence-transformers/all-MiniLM-L6-v2"
model_fname = "mem_model.pkl"

old_model_path = os.path.join(old_model_dir, model_name, model_fname)
new_model_path = os.path.join(new_model_dir, model_name, model_fname)

old_model = joblib.load(old_model_path)
new_model = joblib.load(new_model_path)

assert np.allclose(old_model.coef_, old_model.coef_)
assert np.isclose(new_model.intercept_, new_model.intercept_)
```

ryantwolf reviewed (commented):
Two minor things remain that you might want to fix, but other than that looks good to me.

ryantwolf commented on crossfit/backend/torch/hf/model.py:
```diff
@@ -13,30 +13,57 @@
+        self.start_seq_len = start_seq_len
+        self.seq_len_increment = seq_len_increment
 
+        model = self.load_model("cuda") if not training else None
```
One more thing I thought of actually. Loading the model can take a noticeable amount of time (like 20-30 seconds for Llama 7B on my machine). There is this line of code that seems to check to see if the memory info already exists before running the tests:
```python
    if os.path.exists(mem_model_path):
        return joblib.load(mem_model_path)
```

What do you think about moving the `load_model` call after this early return statement? It should save a bit of time. Otherwise, I've noticed that the model gets loaded for the memory test that doesn't happen, and then it just gets reloaded again on all the workers for the inference which wastes a bit of time.

VibhuJawa replied:
Done.

sarahyurick reviewed (approved):
Tests LGTM, thanks!

VibhuJawa commented:
@ryantwolf , Addressed your review and added improvements, thanks again for the careful review and helping make the performance and user experience better.

ryantwolf commented on crossfit/backend/torch/hf/model.py:
```diff
@@ -19,24 +19,67 @@
+        self.seq_len_increment = seq_len_increment
+
+        cache_dir = os.path.join(
+            CF_HOME, "memory", AutoConfig.from_pretrained(path_or_name)._name_or_path
```
Looks like a small bug was introduced here. This used to call `self.load_cfg` instead of using `AutoConfig`, and it should go back to calling it. During my latest tests this is causing an error because for my config I need to pass in a user access token.

VibhuJawa replied:
Resolved by  https://github.com/rapidsai/crossfit/pull/66/commits/15e27bb33c8a2d56fc8c165d45218c8c902424c0 and https://github.com/rapidsai/crossfit/pull/66/commits/15e27bb33c8a2d56fc8c165d45218c8c902424c0

ryantwolf commented on crossfit/backend/torch/hf/memory_curve_utils.py:
```diff
@@ -0,0 +1,93 @@
+    X: list[list[int]] = []
+    y: list[float] = []
+
+    max_seq = min(AutoTokenizer.from_pretrained(path_or_name).model_max_length, end_seq_len)
```
Similar to the other issue, I think every time `AutoTokenizer` or `AutoConfig` is used in this file it should be using the corresponding methods of the model. Now I'm getting this error with llama guard:
```
Traceback (most recent call last):
  File "/usr/local/bin/aegis_classifier_inference", line 8, in <module>
    sys.exit(console_script())
  File "/usr/local/lib/python3.10/dist-packages/nemo_curator/scripts/aegis_classifier_inference.py", line 126, in console_script
    main()
  File "/usr/local/lib/python3.10/dist-packages/nemo_curator/scripts/aegis_classifier_inference.py", line 62, in main
    domain_classifier = AegisClassifier(
  File "/usr/local/lib/python3.10/dist-packages/nemo_curator/classifiers/aegis.py", line 161, in __init__
    model = AegisHFModel(config=config)
  File "/usr/local/lib/python3.10/dist-packages/nemo_curator/classifiers/aegis.py", line 90, in __init__
    super().__init__(
  File "/usr/local/lib/python3.10/dist-packages/crossfit/backend/torch/hf/model.py", line 58, in __init__
    self.mem = fit_memory_estimate_curve(
  File "/usr/local/lib/python3.10/dist-packages/crossfit/backend/torch/hf/memory_curve_utils.py", line 44, in fit_memory_estimate_curve
    max_seq = min(AutoTokenizer.from_pretrained(path_or_name).model_max_length, end_seq_len)
  File "/usr/local/lib/python3.10/dist-packages/transformers/models/auto/tokenization_auto.py", line 843, in from_pretrained
    return tokenizer_class_fast.from_pretrained(pretrained_model_name_or_path, *inputs, **kwargs)
  File "/usr/local/lib/python3.10/dist-packages/transformers/tokenization_utils_base.py", line 2032, in from_pretrained
    raise EnvironmentError(
OSError: Can't load tokenizer for 'meta-llama/LlamaGuard-7b'. If you were trying to load it from 'https://huggingface.co/models', make sure you don't have a local directory with the same name. Otherwise, make sure 'meta-llama/LlamaGuard-7b' is the correct path to a directory containing all relevant files for a LlamaTokenizerFast tokenizer.
```

VibhuJawa replied:
Resolved by: https://github.com/rapidsai/crossfit/pull/66/commits/987836f8aa52b596bd053f33af240573d50fe267

ryantwolf reviewed (approved):
My tests ran successfully. Thanks so much for doing this @VibhuJawa!
